src/server/pkg/obj/util.go

```diff
@@ -28,3 +30,21 @@ func WithLocalClient(f func(objC Client) error) (retErr error) {
 	}
 	return f(objC)
 }
+
+// Copy copys an object from src at srcPath to dst at dstPath
+func Copy(ctx context.Context, src, dst Client, srcPath, dstPath string) error {
+	rc, err := src.Reader(ctx, srcPath, 0, 0)
+	if err != nil {
+		return err
+	}
+	defer rc.Close()
```

Comment:
I think we should capture error state from the reader / writer close. So, they would look like this:
```
defer func() {
  if err := rc.Close(); retErr == nil {
    retErr = err
  }
}()
```
Also, this would allow you to just return the err from the `io.Copy`, rather than close. I am a bit unsure about whether closing the writer makes sense when an error occurs. I think there was a similar concern when I was working on closing the index writer, and what I went with was always closing with the expectation that the client writer would handle errors that occur while copying (so closing shouldn't store an incomplete state). Here is the code for reference: https://github.com/pachyderm/pachyderm/blob/e5b5944fdc533b6b98516e38d33d17f38193b2a2/src/server/pkg/storage/fileset/index/writer.go#L157

Reply:
This makes sense. Notably wc does not have an "abort" alternative to `Close`. So if the reader errors during the copy, we still call `Close` even though we haven't copied all the data.  The client will get an error and can retry, I suppose. We do have to close the writer though, it can't be left open. We opened it, and if it is backed by the `LimitedClient` then it is consuming one of the allowed active writer slots.